src/string_with_vars.mli

```diff
@@ -6,25 +6,71 @@
 open Import
 
 type t
+(** A sequence of text and variables. *)
+
 val t : t Sexp.Of_sexp.t
-val sexp_of_t : t -> Sexp.t
+(** [t ast] takes an [ast] sexp and returns a string-with-vars.  This
+   function distinguishes between unquoted variables — such as ${@} —
+   and quoted variables — such as "${@}". *)
 
 val loc : t -> Loc.t
+(** [loc t] returns the location of [t] — typically, in the jbuild file. *)
 
-val to_string : t -> string
+val sexp_of_t : t -> Sexp.t
 
-(** [t] generated by the OCaml code. The first argument should be [__POS__]. The second is
-    either a string to parse, a variable name or plain text. *)
-val virt      : (string * int * int * int) -> string -> t
-val virt_var  : (string * int * int * int) -> string -> t
-val virt_text : (string * int * int * int) -> string -> t
+val to_string : t -> string
 
-val just_a_var : t -> string option
+(** [t] generated by the OCaml code. The first argument should be
+   [__POS__]. The second is either a string to parse, a variable name
+   or plain text.  [quoted] says whether the string is quoted ([false]
+   by default). *)
+val virt       : ?quoted: bool -> (string * int * int * int) -> string -> t
+val virt_var   : ?quoted: bool -> (string * int * int * int) -> string -> t
+val virt_text  : (string * int * int * int) -> string -> t
 
 val vars : t -> String_set.t
+(** [vars t] returns the set of all variables in [t]. *)
 
 val fold : t -> init:'a -> f:('a -> Loc.t -> string -> 'a) -> 'a
+(** [fold t ~init ~f] fold [f] on all variables of [t], the text
+   portions being ignored. *)
+
 val iter : t -> f:(Loc.t -> string -> unit) -> unit
+(** [iter t ~f] iterates [f] over all variables of [t], the text
+   portions being ignored. *)
+
+val expand_generic :
```

Comment:
What about putting this and `partial_expand_generic` under a functor? That seem more natural to me.

Reply:
A functor is nice but then you would evaluate it each time for `Action.Unexpanded.Partial.E.expand` to take into account the `~dir` argument.  Of course I can add the `~dir` argument to the functor generated expansion functions but this is not nice...

Reply:
Ah indeed. Ok, let's leave it as it then

Reply:
BTW, why `~dir` is not used [here](https://github.com/ocaml/dune/blob/master/src/super_context.ml#L56) to convert the path?

Reply:
That's likely a mistake. I guess it didn't show up because all these variables point to absolute paths, but we should change it indeed

Reply:
> Ah indeed. Ok, let's leave it as it then

One possibility is to add a "context" for string conversion, context that will need to be passed to the expand functions.  What do you think?

Reply:
Seems good to me

Reply:
You can see the result in the commits below.  I'll attack `Usexp` now.